File: FB_difference.cpp
#include <iostream>
#include <cmath>
using namespace std;

double x_function (double x) {
    return 3*x*x*x -26;
}

void forward (double x, double h, double (*func)(double)){
    cout << "The forward difference derivative is " << (func(x+h) - func(x))/h <<endl;
}

void backward (double x, double h, double (*func)(double)){
    cout << "The backward difference derivative is " << (func(x) - func(x-h))/h <<endl;
}

void central (double x, double h, double (*func)(double)){
    cout << "The central difference derivative is " << ((func(x+h/2) - func(x-h/2))/h)<<endl;
}

void extrapolated(double x, double h, double (*func)(double)){
    double result = (8*func(x+h/4)-8*func(x-h/4)-func(x+h/2)+func(x-h/2))/(3*h);
    cout<< "The extrapolated difference derivative is " << result <<endl;
}

void second (double x, double h, double (*func)(double)){
    double result = 4*(func(x + h/2) + func(x- h/2)- 2*func(x))/(h*h);
    cout << "The second derivative is " <<result<<endl;
}

int main() {
    double x, h;
    int deg;
    cout << "Enter x value: ";
    cin >> x;
    h = 0.0001;

    forward(x,h,x_function);
    backward(x,h,x_function);
    central(x,h,x_function);
    extrapolated(x,h,x_function);
    second(x,h,x_function);
    return 0;
}

File: FB_difference.py
def f(x):
    return 3*x*x*x - 26

def forward (f,x,h):
    return (f(x+h)-f(x))/h

def backward (f,x,h):
    return (f(x)-f(x-h))/h

def central (f,x,h):
    return (f(x+h/2)-f(x-h/2))/h;

def extrapolated (f,x,h):
    return (8*f(x+h/4)-8*f(x-h/4)-f(x+h/2)+f(x-h/2))/(3*h)

def second (f,x,h):
    return 4*(f(x + h/2) + f(x- h/2)- 2*f(x))/(h*h)

x = float(input("Enter x value: "))
acc = int(input("Enter degree of accuracy: "))
h = 10**-acc
print(h)

print("The forward difference derivative is", forward(f,x,h))
print("The backward difference derivative is", backward(f,x,h))
print("The central difference derivative is", central(f,x,h))
print("The extrapolated difference derivative is", extrapolated(f,x,h))
print("The second derivative is", second(f,x,h))
